mlos_bench/mlos_bench/tests/storage/exp_load_test.py

```diff
@@ -0,0 +1,105 @@
+#
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+#
+"""
+Unit tests for the storage subsystem.
+"""
+
+from mlos_bench.storage import Storage
+from mlos_bench.tunables import TunableGroups
+from mlos_bench.environment import Status
+
+# pylint: disable=redefined-outer-name
+
+
+def test_exp_load_empty(exp_storage_memory_sql: Storage.Experiment):
+    """
+    Try to retrieve old experimental data from the empty storage.
+    """
+    (configs, scores) = exp_storage_memory_sql.load()
+    assert not configs
+    assert not scores
+
+
+def test_exp_pending_empty(exp_storage_memory_sql: Storage.Experiment):
+    """
+    Try to retrieve pending experiments from the empty storage.
+    """
+    trials = list(exp_storage_memory_sql.pending())
+    assert not trials
+
+
+def test_exp_trial_pending(exp_storage_memory_sql: Storage.Experiment,
+                           tunable_groups: TunableGroups):
+    """
+    Start a trial and check that it is pending.
+    """
+    trial = exp_storage_memory_sql.trial(tunable_groups)
+    (pending,) = list(exp_storage_memory_sql.pending())
+    assert pending.trial_id == trial.trial_id
+    assert pending.tunables == tunable_groups
+
+
+def test_exp_trial_pending_many(exp_storage_memory_sql: Storage.Experiment,
+                                tunable_groups: TunableGroups):
+    """
+    Start THREE trials and check that both are pending.
+    """
+    config1 = tunable_groups.copy().assign({'rootfs': 'ext4'})
+    config2 = tunable_groups.copy().assign({'rootfs': 'ext2'})
+    trial_ids = {
+        exp_storage_memory_sql.trial(config1).trial_id,
+        exp_storage_memory_sql.trial(config2).trial_id,
+        exp_storage_memory_sql.trial(config2).trial_id,  # Submit same config twice
+    }
+    pending_ids = {pending.trial_id for pending in exp_storage_memory_sql.pending()}
+    assert len(pending_ids) == 3
+    assert trial_ids == pending_ids
+
+
+def test_exp_trial_pending_fail(exp_storage_memory_sql: Storage.Experiment,
+                                tunable_groups: TunableGroups):
+    """
+    Start a trial, fail it, and and check that it is NOT pending.
+    """
+    trial = exp_storage_memory_sql.trial(tunable_groups)
+    trial.update(Status.FAILED)
```

Comment:
New test to make sure we can't update a trial after it's already been updated to non-PENDING?

Reply:
Maybe also add that to the clauses in the `update()` SQL?